File: cc/035.cc
// Project Euler Problem 035
//
// How many circular primes are there below one million?

#include <cstdio>
#include <cmath>
#include "primes1.h"

bool circular(int p, Primes1<int>& primes) {
  if (p<10) return true;
  int len = log10(p);
  int shift = pow(10.0, len);
  for (int i=0; i<len; i++) {
    p = (p % 10) * shift + (p/10);
    if (!primes.is_prime(p)) return false;
  }
  return true;
}

int main(int argc, const char* argv[]) {
  Primes1<int> primes;
  int count = 0;
  int p;
  while ((p=primes.next()) < 1e6) {
    if (circular(p, primes)) {
      count++;
    }
  }
  printf("%d\n", count);
}

File: go/problems/035.go
// Project Euler Problem 035
//
// How many circular primes are there below one million?

package problems

import (
	"fmt"
	"math"

	"../primes"
)

func allRotationsPrime(n int) bool {
	for nn := n / 10; nn > 0; nn /= 10 {
		if nn&1 == 0 {
			return false
		}
	}
	l := int(math.Log10(float64(n)))
	mult := int(math.Pow10(l))
	for i := 0; i < l; i++ {
		n = n/10 + mult*(n%10)
		if !primes.IsPrime(n) {
			return false
		}
	}
	return true
}

func Problem035() string {
	count := 0
	for n := 0; ; n++ {
		if p := primes.Get(n); p < 1e6 {
			if allRotationsPrime(p) {
				count++
			}
		} else {
			break
		}
	}
	return fmt.Sprintf("%d", count)
}

func init() {
	Functions[35] = Problem035
}
